README.md
# apache-hardening (Chef cookbook)

[![Supermarket](http://img.shields.io/cookbook/v/apache-hardening.svg)][1]
[![Build Status](http://img.shields.io/travis/dev-sec/chef-apache-hardening.svg)][2]
[![Code Coverage](http://img.shields.io/coveralls/dev-sec/chef-apache-hardening.svg)][3]
[![Dependencies](http://img.shields.io/gemnasium/dev-sec/chef-apache-hardening.svg)][4]
[![Gitter Chat](https://badges.gitter.im/Join%20Chat.svg)][5]

## Description

This cookbook provides a secure overlay for apache configuration.

## Requirements

* chef

### Platform

- Debian 7, 8
- Ubuntu 14.04, 16.04
- CentOS 6.8, 7.2
- OracleLinux 6.8, 7.2

## Usage

A sample role may look like:

```json
{
    "name": "apache",
    "default_attributes": { },
    "override_attributes": { },
    "json_class": "Chef::Role",
    "description": "Apache Hardened Server Test Role",
    "chef_type": "role",
    "run_list": [
        "recipe[apt]",
        "recipe[apache2]",
        "recipe[apache-hardening]"
    ]
}
```

## Recipes

### apache-hardening::hardening (default)

This recipe is an overlay recipe for the [apache2 cookbook](https://github.com/viverae-cookbooks/apache2) and applies `apache-hardening::hardening`

Add the following to your runlist and customize security option attributes

```
"recipe[apache2]",
"recipe[apache-hardening]"
```

This hardening recipe installs the hardening but expects an existing installation of Apache2.

## Security Options

* `node['apache']['traceenable'] = 'Off'`
   This directive overrides the behavior of TRACE for both the core server and mod_proxy.
   See [http://httpd.apache.org/docs/2.2/mod/core.html#traceenable](http://httpd.apache.org/docs/2.2/mod/core.html#traceenable) for details
   Defaults to: `Off`

* `node['apache_hardening']['allowed_http_methods'] = %w( GET POST )`
   A list of HTTP methods that should be allowed in the server.
   See [http://httpd.apache.org/docs/trunk/mod/mod_allowmethods.html](http://httpd.apache.org/docs/trunk/mod/mod_allowmethods.html) for details
   Defaults to: `GET POST`

* `node['apache_hardening']['modules_to_disable'] = %w( cgi cgid )`
   This parameter sets a list of modules that should be disabled on the target server.
   See [http://httpd.apache.org/docs/current/mod/](http://httpd.apache.org/docs/current/mod/) for details
   Defaults to: `cgi cgid`

## Tests

For local testing you can use vagrant or docker to run tests locally. You will have to install Virtualbox and Vagrant or docker on your system. See [Vagrant Downloads](http://downloads.vagrantup.com/) for a vagrant or [Docker Downloads](https://docs.docker.com/mac/) package suitable for your system. For all our tests we use `test-kitchen`. If you are not familiar with `test-kitchen` please have a look at [their guide](http://kitchen.ci/docs/getting-started).

```
# Install dependencies
gem install bundler
bundle install

# Do lint checks
bundle exec rake lint

# fast test on one machine
bundle exec kitchen test default-ubuntu-1404

# test on all machines
bundle exec kitchen test

# for development
bundle exec kitchen create default-ubuntu-1404
bundle exec kitchen converge default-ubuntu-1404
bundle exec kitchen verify default-ubuntu-1204
```

## Contributors + Kudos

* Dominik Richter [arlimus](https://github.com/arlimus)
* Christoph Hartmann [chris-rock](https://github.com/chris-rock)
* Patrick Muench [atomic111](https://github.com/atomic111)
* Edmund Haselwanter [ehaselwanter](https://github.com/ehaselwanter)

## Contributing

See [contributor guideline](CONTRIBUTING.md).

## License and Author

* Author:: Deutsche Telekom AG

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[1]: https://supermarket.getchef.com/cookbooks/apache-hardening
[2]: http://travis-ci.org/dev-sec/chef-apache-hardening
[3]: https://coveralls.io/r/dev-sec/chef-apache-hardening
[4]: https://gemnasium.com/dev-sec/chef-apache-hardening
[5]: https://gitter.im/dev-sec/general


CHANGELOG.md
# Change Log

## [1.0.0](https://github.com/dev-sec/chef-apache-hardening/tree/1.0.0) (2016-06-24)
**Closed issues:**

- Bad twiddle-wakka in metadata [\#19](https://github.com/dev-sec/chef-apache-hardening/issues/19)

**Merged pull requests:**

- switch to docker and use inspec for testing [\#24](https://github.com/dev-sec/chef-apache-hardening/pull/24) ([atomic111](https://github.com/atomic111))
- upgrade to Berkshelf 4 [\#23](https://github.com/dev-sec/chef-apache-hardening/pull/23) ([chris-rock](https://github.com/chris-rock))
- update common kitchen.yml platforms [\#22](https://github.com/dev-sec/chef-apache-hardening/pull/22) ([chris-rock](https://github.com/chris-rock))
- Added a missing dash [\#21](https://github.com/dev-sec/chef-apache-hardening/pull/21) ([efreesen](https://github.com/efreesen))
- closes \#19 [\#20](https://github.com/dev-sec/chef-apache-hardening/pull/20) ([chris-rock](https://github.com/chris-rock))
- update common Gemfile for chef11+12 [\#18](https://github.com/dev-sec/chef-apache-hardening/pull/18) ([arlimus](https://github.com/arlimus))
- common files: centos7 + rubocop [\#17](https://github.com/dev-sec/chef-apache-hardening/pull/17) ([arlimus](https://github.com/arlimus))
- update common kitchen.yml platforms [\#16](https://github.com/dev-sec/chef-apache-hardening/pull/16) ([arlimus](https://github.com/arlimus))
- update common readme badges [\#15](https://github.com/dev-sec/chef-apache-hardening/pull/15) ([arlimus](https://github.com/arlimus))
- updating common files [\#11](https://github.com/dev-sec/chef-apache-hardening/pull/11) ([arlimus](https://github.com/arlimus))
- update readme and add contributing section [\#10](https://github.com/dev-sec/chef-apache-hardening/pull/10) ([ehaselwanter](https://github.com/ehaselwanter))
- add some more specs for the hardening code [\#9](https://github.com/dev-sec/chef-apache-hardening/pull/9) ([ehaselwanter](https://github.com/ehaselwanter))
- add badges to readme [\#8](https://github.com/dev-sec/chef-apache-hardening/pull/8) ([chris-rock](https://github.com/chris-rock))
- add specs to apache cookbook [\#7](https://github.com/dev-sec/chef-apache-hardening/pull/7) ([ehaselwanter](https://github.com/ehaselwanter))
- fix alias conf according to gis requirements [\#6](https://github.com/dev-sec/chef-apache-hardening/pull/6) ([ehaselwanter](https://github.com/ehaselwanter))
- rubocop 0.27 is picky about the Berksfile [\#5](https://github.com/dev-sec/chef-apache-hardening/pull/5) ([ehaselwanter](https://github.com/ehaselwanter))
- remove o-rwx [\#4](https://github.com/dev-sec/chef-apache-hardening/pull/4) ([ehaselwanter](https://github.com/ehaselwanter))
- Update common [\#3](https://github.com/dev-sec/chef-apache-hardening/pull/3) ([arlimus](https://github.com/arlimus))
- fix reload issue. we need a restart action [\#2](https://github.com/dev-sec/chef-apache-hardening/pull/2) ([ehaselwanter](https://github.com/ehaselwanter))
- add the chef hardeing config [\#1](https://github.com/dev-sec/chef-apache-hardening/pull/1) ([ehaselwanter](https://github.com/ehaselwanter))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

spec/spec_helper.rb
# encoding: utf-8

require 'chefspec'
require 'chefspec/berkshelf'
require 'coveralls'

# coverage report
Coveralls.wear!
at_exit { ChefSpec::Coverage.report! }


spec/unit/recipes/default_spec.rb
# encoding: utf-8

require 'spec_helper'

describe 'apache-hardening::default' do

  before { allow_any_instance_of(Chef::Recipe).to receive(:search) }
  let(:runner) { ChefSpec::ServerRunner.new }
  let(:node) { runner.node }
  let(:chef_run) { runner.converge(described_recipe) }

  before do
    stub_command('/usr/sbin/apache2 -t')
    stub_command("find /etc/apache2 -perm -o+r -type f -o -perm -o+w -type f | wc -l | egrep '^0$'")
  end

  it 'includes apache-hardening::hardening recipe' do
    expect(chef_run).to include_recipe('apache-hardening::hardening')
  end

end


spec/unit/recipes/hardening_spec.rb
# encoding: utf-8

require 'spec_helper'

describe 'apache-hardening::hardening' do

  before { allow_any_instance_of(Chef::Recipe).to receive(:search) }
  let(:runner) { ChefSpec::ServerRunner.new }
  let(:node) { runner.node }
  let(:chef_run) { runner.converge(described_recipe) }

  before do
    stub_command('/usr/sbin/apache2 -t')
    stub_command("find /etc/apache2 -perm -o+r -type f -o -perm -o+w -type f | wc -l | egrep '^0$'")
  end

  it 'creates hardening.conf with correct permissions' do

    expect(chef_run).to create_template(File.join(chef_run.node['apache']['dir'], '/conf-enabled/', 'hardening.conf')).with(
      user:   'root',
      group:  chef_run.node['apache']['root_group'],
      mode:   '0640'
    )

  end

  # Since we cannot test the definition apache_module itself, we have to test for a side effect to happen
  # We check for the not present module file in mods_enabled
  it 'disables specified apache modules' do

    chef_run.node['apache_hardening']['modules_to_disable'].each do |module_to_disable|
      expect(chef_run).to_not create_file_if_missing("#{chef_run.node['apache']['dir']}/mods-enabled/#{module_to_disable}.load")
    end

  end

  it 'checks that the correct alias.conf.erb template is being used' do

    chef_run.run_context.resource_collection.each do |resource|

      next unless resource.name == "#{chef_run.node['apache']['dir']}/mods-available/alias.conf"

      expect(resource.cookbook).to eq('apache-hardening')

    end

  end

  it 'makes sure that it does not flap on o-rw' do

    chef_run.run_context.resource_collection.each do |resource|

      next unless (resource.respond_to? :mode) && !resource.mode.nil?

      expect(resource.mode).to eq('0750') if resource.name =~ /#{chef_run.node['apache']['dir']}/ && resource.mode[1].eql?('7')
      expect(resource.mode).to eq('0640') if resource.name =~ /#{chef_run.node['apache']['dir']}/ && resource.mode[1].eql?('6')
      expect(resource.mode).to eq('0640') if resource.name == 'apache2.conf'

    end

  end

  it 'executes "remove world readable files"' do

    stub_command("find /etc/apache2 -perm -o+r -type f -o -perm -o+w -type f | wc -l | egrep '^0$'").and_return(false)
    expect(chef_run).to run_execute('remove world readable files')

  end

  it 'does not execute "remove world readable files"' do

    stub_command("find /etc/apache2 -perm -o+r -type f -o -perm -o+w -type f | wc -l | egrep '^0$'").and_return(true)
    expect(chef_run).to_not run_execute('remove world readable files')

  end

end


recipes/hardening.rb
# encoding: utf-8
#
# Cookbook Name: apache-hardening
# Recipe: hardening.rb
#
# Copyright 2014, Edmund Haselwanter
# Copyright 2014, Deutsche Telekom AG
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

template File.join(node['apache']['dir'], '/conf-enabled/', 'hardening.conf') do
  action :create
  source 'hardening.cnf.erb'
  owner 'root'
  group node['apache']['root_group']
  mode '0640'
  notifies :restart, 'service[apache2]', :delayed
end

node['apache_hardening']['modules_to_disable'].each do |module_to_disable|
  apache_module module_to_disable do
    enable false
    notifies :restart, 'service[apache2]', :delayed
  end
end

begin
  r = resources(template: "#{node['apache']['dir']}/mods-available/alias.conf")
  r.cookbook('apache-hardening')
rescue Chef::Exceptions::ResourceNotFound
  Chef::Log.info("ignoring update of alias.conf since it is not used #{node['apache']['dir']}/mods-available/alias.conf")
end

# change all the already created resource so we do not flap on o-rw
run_context.resource_collection.each do |resource|
  resource.mode('0640') if resource.name =~ /#{node['apache']['dir']}/ && resource.mode == '0644'
  resource.mode('0750') if resource.name =~ /#{node['apache']['dir']}/ && resource.mode == '0755'
  resource.mode('0640') if resource.name == 'apache2.conf'
end

# change all the other files not defined as resources
execute 'remove world readable files' do
  command "chmod -R o-rw #{node['apache']['dir']}"
  not_if "find #{node['apache']['dir']} -perm -o+r -type f -o -perm -o+w -type f | wc -l | egrep '^0$'"
end
